### DISCUSSION TITLE
What is best way to specify optional fields in pydantic?

### DISCUSSION THREAD

**Discussion by alvassin** [pydantic/pydantic#2462] [Question] [answered]
I see three different ways, that have same effect (for the first point of view).

```
a: Optional[str]
b: str = None
c: Optional[str] = None
```

What is the correct one? Is there any difference?

**Comment by PrettyWood** [comment, marked as answer]
Hi @alvassin
Currently it's the same but it will probably be changed in v2
Quoting proposal from #2353
```py
class Model(BaseModel):
    required: str
    nullable_required: Optional[str]

    optional_with_default_value: str = 'pika'
    optional_nullable_with_default_value: Optional[str] = None
    optional_without_default_value: str = Field(omit_default=True)  # default is `Undefined` (already the case in `Field()`)
    optional_nullable_without_default_value: Optional[str] = Field(omit_default=True)

    # INVALID as `None` is not valid for type `str`
    invalid_optional: str = None
```
In your case I would go with either `b` or `c` for now (a field with a default value)

**Reply by ricosaurus** [reply]
Will the same work for BaseSettings rather than BaseModel?
I am currently converting my standard dataclasses to pydantic models, and have relied on the 'Unset' singleton pattern to give values to attributes that are required with known types but unknown values at model initiation -- avoiding the None confusion, and allowing me to later check all fields for Unset, regardless of whether None is a valid value, before they are needed.
But I just found that this singleton works inheriting from BaseModel, but not from BaseSettings, which I primarily need:
```

from pydantic import BaseModel, BaseSettings, Field

class _UNSET_TYPE:
    
    def __repr__(self):
        return "UNSET_TYPE"

UNSET = _UNSET_TYPE()

class FooModel(BaseModel):
    
    bar: int = Field(default=UNSET)

class FooSettings(BaseSettings):
    
    bar: int = Field(default=UNSET)
        
fm = FooModel()
print(fm.bar)

fs = FooSettings()
```
which produces:

> UNSET_TYPE
> 
> ---------------------------------------------------------------------------
> ValidationError                           Traceback (most recent call last)
> <ipython-input-289-dbcafa6f2974> in <module>
>      19 print(fm.bar)
>      20 
> ---> 21 fs = FooSettings()
> 
> ~/anaconda3/envs/py39/lib/python3.9/site-packages/pydantic/env_settings.cpython-39-x86_64-linux-gnu.so in pydantic.env_settings.BaseSettings.__init__()
> 
> ~/anaconda3/envs/py39/lib/python3.9/site-packages/pydantic/main.cpython-39-x86_64-linux-gnu.so in pydantic.main.BaseModel.__init__()
> 
> ValidationError: 1 validation error for FooSettings
> bar
>   value is not a valid integer (type=type_error.integer)

###
Is there currently a work around for BaseSettings?
I am new to pydantic and still learning the ropes.  And thinking ahead, come v2 I will be able to use your 3rd case there -- 'optional_without_default_value' with either BaseModel or BaseSettings?

Thanks for the clarity -- was getting rather lost in the optional requireds in the Issues. ;-)

**Reply by PrettyWood** [reply]
`BaseSettings` sets `Config.validate_all` to `True` by default (whereas `BaseModel` sets it to `False`).
This is why `BaseSettings` validates also default values and `UNSET` is indeed not a valid `int`.
Two solutions:
```py
class FooSettings(BaseSettings, validate_all=False):  # force `Config.validate_all = False`
    bar: int = UNSET
```
or
```py
# create a proper singleton that is compliants with mypy: https://www.python.org/dev/peps/pep-0484/#id30
class Unset(Enum):
    token = 0

    def __repr__(self) -> str:
        return 'UNSET'

    def __str__(self) -> str:
        return 'UNSET'

UNSET = Unset.token

# add `Unset` in allowed types when you set `UNSET`
class FooSettings(BaseSettings):
    bar: Union[int, Unset] = UNSET

fs = FooSettings()
print(fs.bar)
# 'UNSET'
```

**Reply by ricosaurus** [reply]
Thanks much.  Was aware of neither of these.

**Comment by mattixpet** [comment]
Just to add to PrettyWood's excellent example, I didn't get the desired behavior until I did

```python
optional_nullable_without_default_value: Optional[str] = Field("Never a real value", exclude_if=lambda x: x == "Never a real value")
```

Although that is perhaps a FastAPI issue, not pydantic.

I'm still baffled how complicated this is to do and find info on, as opposed to typescripts simple `key?` syntax. I must be doing something wrong..
